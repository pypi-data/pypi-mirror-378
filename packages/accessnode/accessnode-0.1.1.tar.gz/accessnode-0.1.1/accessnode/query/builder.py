# from typing import Any, List, Dict, Optional, TypeVar, Generic, Type
# from .types import (
#     JoinType, OrderDirection, Operator, WhereClause,
#     JoinClause, OrderByClause, QueryValue
# )
# from ..models.base import BaseModel
# from ..core.exceptions import QueryError

# T = TypeVar('T', bound=BaseModel)

# class QueryBuilder(Generic[T]):
#     def __init__(self, model_class: Type[T]):
#         self.model = model_class
#         self.selected_fields: List[str] = []
#         self.where_clauses: List[WhereClause] = []
#         self.join_clauses: List[JoinClause] = []
#         self.order_by_clauses: List[OrderByClause] = []
#         self.group_by_fields: List[str] = []
#         self.having_clauses: List[WhereClause] = []
#         self._limit: Optional[int] = None
#         self._offset: Optional[int] = None
#         self._distinct: bool = False

#     def select(self, *fields: str) -> 'QueryBuilder[T]':
#         """Select specific fields."""
#         self.selected_fields.extend(fields)
#         return self

#     def distinct(self) -> 'QueryBuilder[T]':
#         """Make the query distinct."""
#         self._distinct = True
#         return self

#     def where(
#         self,
#         field: str,
#         operator: Operator,
#         value: QueryValue
#     ) -> 'QueryBuilder[T]':
#         """Add WHERE clause."""
#         self.where_clauses.append(
#             WhereClause(field, operator, value)
#         )
#         return self

#     def or_where(
#         self,
#         field: str,
#         operator: Operator,
#         value: QueryValue
#     ) -> 'QueryBuilder[T]':
#         """Add OR WHERE clause."""
#         self.where_clauses.append(
#             WhereClause(field, operator, value, is_or=True)
#         )
#         return self

#     def where_in(self, field: str, values: List[Any]) -> 'QueryBuilder[T]':
#         """Add WHERE IN clause."""
#         return self.where(field, Operator.IN, values)

#     def where_not_in(self, field: str, values: List[Any]) -> 'QueryBuilder[T]':
#         """Add WHERE NOT IN clause."""
#         return self.where(field, Operator.NOT_IN, values)

#     def where_null(self, field: str) -> 'QueryBuilder[T]':
#         """Add WHERE IS NULL clause."""
#         return self.where(field, Operator.IS_NULL, None)

#     def where_not_null(self, field: str) -> 'QueryBuilder[T]':
#         """Add WHERE IS NOT NULL clause."""
#         return self.where(field, Operator.IS_NOT_NULL, None)

#     def where_between(
#         self,
#         field: str,
#         start: Any,
#         end: Any
#     ) -> 'QueryBuilder[T]':
#         """Add WHERE BETWEEN clause."""
#         return self.where(field, Operator.BETWEEN, [start, end])

#     def join(
#         self,
#         table: str,
#         field: str,
#         operator: Operator,
#         value: str,
#         type: JoinType = JoinType.INNER
#     ) -> 'QueryBuilder[T]':
#         """Add JOIN clause."""
#         self.join_clauses.append(
#             JoinClause(
#                 table,
#                 type,
#                 [WhereClause(field, operator, value)]
#             )
#         )
#         return self

#     def left_join(
#         self,
#         table: str,
#         field: str,
#         operator: Operator,
#         value: str
#     ) -> 'QueryBuilder[T]':
#         """Add LEFT JOIN clause."""
#         return self.join(table, field, operator, value, JoinType.LEFT)

#     def right_join(
#         self,
#         table: str,
#         field: str,
#         operator: Operator,
#         value: str
#     ) -> 'QueryBuilder[T]':
#         """Add RIGHT JOIN clause."""
#         return self.join(table, field, operator, value, JoinType.RIGHT)

#     def order_by(
#         self,
#         field: str,
#         direction: OrderDirection = OrderDirection.ASC
#     ) -> 'QueryBuilder[T]':
#         """Add ORDER BY clause."""
#         self.order_by_clauses.append(OrderByClause(field, direction))
#         return self

#     def group_by(self, *fields: str) -> 'QueryBuilder[T]':
#         """Add GROUP BY clause."""
#         self.group_by_fields.extend(fields)
#         return self

#     def having(
#         self,
#         field: str,
#         operator: Operator,
#         value: QueryValue
#     ) -> 'QueryBuilder[T]':
#         """Add HAVING clause."""
#         self.having_clauses.append(
#             WhereClause(field, operator, value)
#         )
#         return self

#     def limit(self, limit: int) -> 'QueryBuilder[T]':
#         """Set LIMIT clause."""
#         if limit < 0:
#             raise QueryError("Limit must be non-negative")
#         self._limit = limit
#         return self

#     def offset(self, offset: int) -> 'QueryBuilder[T]':
#         """Set OFFSET clause."""
#         if offset < 0:
#             raise QueryError("Offset must be non-negative")
#         self._offset = offset
#         return self

#     def paginate(self, page: int, per_page: int = 20) -> 'QueryBuilder[T]':
#         """Add pagination."""
#         if page < 1:
#             raise QueryError("Page must be positive")
#         if per_page < 1:
#             raise QueryError("Items per page must be positive")
        
#         self._limit = per_page
#         self._offset = (page - 1) * per_page
#         return self

#     async def count(self) -> int:
#         """Get count of records."""
#         original_fields = self.selected_fields.copy()
#         self.selected_fields = ["COUNT(*) as count"]
#         result = await self.first()
#         self.selected_fields = original_fields
#         return result.count if result else 0

#     async def exists(self) -> bool:
#         """Check if any records exist."""
#         return await self.count() > 0

#     async def first(self) -> Optional[T]:
#         """Get first record."""
#         self._limit = 1
#         results = await self.get()
#         return results[0] if results else None

#     async def first_or_fail(self) -> T:
#         """Get first record or raise exception."""
#         result = await self.first()
#         if not result:
#             raise QueryError("No records found")
#         return result

#     async def get(self) -> List[T]:
#         """Execute query and get results."""
#         # This will be implemented by each database adapter
#         pass

#     async def update(self, data: Dict[str, Any]) -> int:
#         """Update records matching query."""
#         # This will be implemented by each database adapter
#         pass

#     async def delete(self) -> int:
#         """Delete records matching query."""
#         # This will be implemented by each database adapter
#         pass

#     def clone(self) -> 'QueryBuilder[T]':
#         """Create a clone of the current query."""
#         new_query = QueryBuilder(self.model)
#         new_query.selected_fields = self.selected_fields.copy()
#         new_query.where_clauses = self.where_clauses.copy()
#         new_query.join_clauses = self.join_clauses.copy()
#         new_query.order_by_clauses = self.order_by_clauses.copy()
#         new_query.group_by_fields = self.group_by_fields.copy()
#         new_query.having_clauses = self.having_clauses.copy()
#         new_query._limit = self._limit
#         new_query._offset = self._offset
#         new_query._distinct = self._distinct
#         return new_query






































# # # from typing import Any, Dict, List, Optional, Union, Type
# # # from ..models.base import BaseModel
# # # from .filters import FilterExpression
# # # from .compiler import QueryCompiler

# # # class QueryBuilder:
# # #     """Builds database queries with type safety"""
    
# # #     def __init__(self, model: Type[BaseModel]):
# # #         self.model = model
# # #         self.compiler = QueryCompiler()
# # #         self._reset()
    
# # #     def _reset(self):
# # #         self.selected_fields = []
# # #         self.where_clauses = []
# # #         self.order_by_clauses = []
# # #         self.includes = []
# # #         self.limit_value = None
# # #         self.offset_value = None
    
# # #     def select(self, *fields: str) -> 'QueryBuilder':
# # #         """Select specific fields"""
# # #         self.selected_fields.extend(fields)
# # #         return self
    
# # #     def where(self, *filters: FilterExpression) -> 'QueryBuilder':
# # #         """Add WHERE clauses"""
# # #         self.where_clauses.extend(filters)
# # #         return self
    
# # #     def order_by(self, field: str, ascending: bool = True) -> 'QueryBuilder':
# # #         """Add ORDER BY clause"""
# # #         self.order_by_clauses.append((field, ascending))
# # #         return self
    
# # #     def include(self, relation: str) -> 'QueryBuilder':
# # #         """Include related models"""
# # #         self.includes.append(relation)
# # #         return self
    
# # #     def limit(self, value: int) -> 'QueryBuilder':
# # #         """Add LIMIT clause"""
# # #         self.limit_value = value
# # #         return self
    
# # #     def offset(self, value: int) -> 'QueryBuilder':
# # #         """Add OFFSET clause"""
# # #         self.offset_value = value
# # #         return self
    
# # #     def build(self) -> str:
# # #         """Build the final query"""
# # #         return self.compiler.compile(self)


# # # # from typing import Any, Dict, List, Optional
# # # # from .operators import Operator

# # # # class QueryBuilder:
# # # #     def __init__(self):
# # # #         self.select_fields: List[str] = []
# # # #         self.where_conditions: List[Dict] = []
# # # #         self.include_relations: Dict[str, Any] = {}
# # # #         self.order_by: List[Dict] = []
# # # #         self.limit_value: Optional[int] = None
# # # #         self.offset_value: Optional[int] = None
# # # #         self.group_by: List[str] = []

# # # #     def select(self, *fields: str) -> 'QueryBuilder':
# # # #         self.select_fields.extend(fields)
# # # #         return self

# # # #     def where(self, field: str, operator: Operator, value: Any) -> 'QueryBuilder':
# # # #         self.where_conditions.append({
# # # #             'field': field,
# # # #             'operator': operator,
# # # #             'value': value
# # # #         })
# # # #         return self

# # # #     def include(self, relation: str, nested: Optional[Dict] = None) -> 'QueryBuilder':
# # # #         self.include_relations[relation] = nested or {}
# # # #         return self

# # # #     def order_by_asc(self, field: str) -> 'QueryBuilder':
# # # #         self.order_by.append({'field': field, 'direction': 'ASC'})
# # # #         return self

# # # #     def order_by_desc(self, field: str) -> 'QueryBuilder':
# # # #         self.order_by.append({'field': field, 'direction': 'DESC'})
# # # #         return self

# # # #     def limit(self, value: int) -> 'QueryBuilder':
# # # #         self.limit_value = value
# # # #         return self

# # # #     def offset(self, value: int) -> 'QueryBuilder':
# # # #         self.offset_value = value
# # # #         return self

# # # #     def group_by(self, *fields: str) -> 'QueryBuilder':
# # # #         self.group_by.extend(fields)
# # # #         return self

# # #  accessnode/query/builder.py
# # from typing import Any, Dict, List, Optional
# # from accessnode.query.base_query import BaseQueryBuilder
# # from accessnode.database.types import QueryOperator

# # class QueryBuilder(BaseQueryBuilder):
# #     def select(self, *fields: str) -> 'QueryBuilder':
# #         """Select specific fields"""
# #         self.selected_fields.extend(fields)
# #         return self

# #     def filter(self, **conditions: Dict[str, Any]) -> 'QueryBuilder':
# #         """Add filter conditions"""
# #         for field, value in conditions.items():
# #             if '__' in field:
# #                 field_name, operator = field.split('__')
# #                 self.filters[field_name] = {
# #                     'operator': QueryOperator(operator),
# #                     'value': value
# #                 }
# #             else:
# #                 self.filters[field] = {
# #                     'operator': QueryOperator.EQUALS,
# #                     'value': value
# #                 }
# #         return self

# #     def limit(self, limit: int) -> 'QueryBuilder':
# #         """Set limit"""
# #         self._limit = limit
# #         return self

# #     def offset(self, offset: int) -> 'QueryBuilder':
# #         """Set offset"""
# #         self._offset = offset
# #         return self

# #     def order_by(self, field: str, descending: bool = False) -> 'QueryBuilder':
# #         """Add order by clause"""
# #         self._order_by.append({
# #             'field': field,
# #             'descending': descending
# #         })
# #         return self

# #     def include(self, relation: str) -> 'QueryBuilder':
# #         """Include related models"""
# #         self._includes.append(relation)
# #         return self

# #     def update(self, data: Dict[str, Any]) -> 'QueryBuilder':
# #         """Set update data"""
# #         self._update_data = data
# #         return self

# #     def delete(self) -> 'QueryBuilder':
# #         """Mark as delete query"""
# #         self._is_delete = True
# #         return self

# #     def insert(self, data: Dict[str, Any]) -> 'QueryBuilder':
# #         """Set insert data"""
# #         self._insert_data = data
# #         return self

# #     async def execute(self) -> List[Dict[str, Any]]:
# #         """Execute the query"""
# #         # This will be implemented by each database adapter
# #         pass

# #     async def count(self) -> int:
# #         """Get count of records"""
# #         # This will be implemented by each database adapter
# #         pass





