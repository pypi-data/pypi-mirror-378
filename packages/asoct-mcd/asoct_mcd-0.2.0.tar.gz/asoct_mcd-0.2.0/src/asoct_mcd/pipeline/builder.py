"""
Author: B. Chen

Builder class for MCD pipeline construction.
"""

import yaml
from typing import Dict, Any

from ..model_management import ModelFactory
from ..segmentation import SegmentorFactory
from ..threshold_methods import ThresholdFactory
from ..classification import ClassifierFactory
from ..prompt_generation import PromptGeneratorFactory
from ..preprocessing import PreprocessingPipeline, DenoiserFactory, ConverterFactory

from .config import MCDPipelineConfig
from .wrappers import (
    ZeroShotSegmentationWrapper, TrainedSegmentationWrapper,
    ClassificationWrapper, PromptGenerationWrapper
)
from .pipeline import MCDPipeline


class MCDPipelineBuilder:
    """Builder for MCD pipeline with configuration support."""
    
    def __init__(self, config: MCDPipelineConfig = None):
        """Initialize builder with optional configuration."""
        self.config = config or MCDPipelineConfig()
        self._segmentor = None
        self._threshold_method = None
        self._classifier = None
        self._preprocessing = None
    
    def from_config(self, config: MCDPipelineConfig) -> 'MCDPipelineBuilder':
        """Set configuration and reset builder."""
        self.config = config
        return self.reset()
    
    def from_dict(self, config_dict: Dict[str, Any]) -> 'MCDPipelineBuilder':
        """Create from dictionary configuration."""
        self.config = MCDPipelineConfig.from_dict(config_dict)
        return self.reset()
    
    def from_yaml(self, yaml_path: str) -> 'MCDPipelineBuilder':
        """Create from YAML configuration file."""
        with open(yaml_path, 'r') as f:
            config_dict = yaml.safe_load(f)
        return self.from_dict(config_dict)
    
    def reset(self) -> 'MCDPipelineBuilder':
        """Reset all components."""
        self._segmentor = None
        self._threshold_method = None
        self._classifier = None
        self._preprocessing = None
        return self
    
    def build(self) -> MCDPipeline:
        """Build configured pipeline."""
        # Auto-build components from config
        self._build_segmentor()
        self._build_threshold_method()
        self._build_classifier()
        if self.config.preprocessing.enable:
            self._build_preprocessing()
        
        # Validate components
        if not all([self._segmentor, self._threshold_method, self._classifier]):
            raise ValueError("Missing required pipeline components")
        
        return MCDPipeline(
            segmentor=self._segmentor,
            threshold_method=self._threshold_method,
            classifier=self._classifier,
            preprocessing_pipeline=self._preprocessing,
            config=self.config
        )
    
    def _build_segmentor(self):
        """Build segmentor from configuration."""
        seg_config = self.config.segmentation
        
        if seg_config.method == "zero_shot":
            # Create model wrapper
            model_wrapper = ModelFactory.create_model(seg_config.model_name)
            segmentation_wrapper = ZeroShotSegmentationWrapper(model_wrapper)
            
            # Create prompt generator
            prompt_gen = PromptGeneratorFactory.create_generator(
                seg_config.prompt_generator,
                offset_ratio=seg_config.offset_ratio,
                area_ratio_threshold=seg_config.area_ratio_threshold
            )
            prompt_wrapper = PromptGenerationWrapper(prompt_gen)
            
            self._segmentor = SegmentorFactory.create_segmentor(
                "zero_shot",
                model_wrapper=segmentation_wrapper,
                prompt_generator=prompt_wrapper
            )
        
        elif seg_config.method == "trained":
            model_wrapper = ModelFactory.create_model(seg_config.model_name)
            segmentation_wrapper = TrainedSegmentationWrapper(model_wrapper)
            
            self._segmentor = SegmentorFactory.create_segmentor(
                "trained",
                model_wrapper=segmentation_wrapper
            )
    
    def _build_threshold_method(self):
        """Build threshold method from configuration."""
        self._threshold_method = ThresholdFactory.create_threshold(
            self.config.threshold.method
        )
    
    def _build_classifier(self):
        """Build classifier from configuration."""
        model_wrapper = ModelFactory.create_model(self.config.classification.model_name)
        classification_wrapper = ClassificationWrapper(model_wrapper)
        
        self._classifier = ClassifierFactory.create_classifier(
            "cell",
            model_wrapper=classification_wrapper
        )
    
    def _build_preprocessing(self):
        """Build preprocessing pipeline from configuration."""
        prep_config = self.config.preprocessing
        pipeline = PreprocessingPipeline()
        
        if prep_config.convert_to_grayscale:
            converter = ConverterFactory.create_converter('grayscale')
            pipeline.add_converter(converter)
        
        if prep_config.denoising_method != "none":
            if prep_config.denoising_method == "nlm":
                denoiser = DenoiserFactory.create_denoiser(
                    'nlm',
                    h=prep_config.nlm_h,
                    template_window=prep_config.nlm_template_window,
                    search_window=prep_config.nlm_search_window
                )
            else:
                denoiser = DenoiserFactory.create_denoiser(prep_config.denoising_method)
            
            pipeline.add_denoiser(denoiser)
        
        self._preprocessing = pipeline