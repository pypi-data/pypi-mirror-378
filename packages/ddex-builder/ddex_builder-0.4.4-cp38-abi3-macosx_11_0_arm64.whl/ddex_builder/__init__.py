"""DDEX Builder - High-performance DDEX XML builder with deterministic output."""

__version__ = "0.4.3"

# Try to import the Rust extension
try:
    from ddex_builder._internal import (
        DdexBuilder as _RustDdexBuilder,
        Release as _RustRelease,
        Resource as _RustResource,
        ValidationResult as _RustValidationResult,
        BuilderStats as _RustBuilderStats,
        BuildResult as _RustBuildResult,
        PresetInfo as _RustPresetInfo,
        FidelityOptions as _RustFidelityOptions,
        VerificationResult as _RustVerificationResult,
        ValidationRulePy as _RustValidationRule,
        batch_build as _rust_batch_build,
        validate_structure as _rust_validate_structure,
    )
    _RUST_AVAILABLE = True

    # Use Rust implementations directly
    DdexBuilder = _RustDdexBuilder
    Release = _RustRelease
    Resource = _RustResource
    ValidationResult = _RustValidationResult
    BuilderStats = _RustBuilderStats
    BuildResult = _RustBuildResult
    PresetInfo = _RustPresetInfo
    FidelityOptions = _RustFidelityOptions
    VerificationResult = _RustVerificationResult
    ValidationRule = _RustValidationRule
    batch_build = _rust_batch_build
    validate_structure = _rust_validate_structure

except ImportError as e:
    # Only fall back to mock if truly not available
    _RUST_AVAILABLE = False
    print(f"Warning: Rust extension not available, using mock implementation: {e}")

    # Mock implementations for development only
    class DdexBuilder:
        """Mock DdexBuilder for development/testing."""
        def __init__(self):
            self._releases = []
            self._resources = []

        def add_release(self, release):
            self._releases.append(release)
            return self

        def add_resource(self, resource):
            self._resources.append(resource)
            return self

        def build(self):
            return f"<mock>Built {len(self._releases)} releases</mock>"

        def validate(self):
            return {"valid": True, "errors": []}

        def from_dataframe(self, df):
            return self

        def get_stats(self):
            return {"releases": len(self._releases), "resources": len(self._resources)}

    class Release:
        """Mock Release for development."""
        def __init__(self):
            self.release_id = None
            self.title = None

    class Resource:
        """Mock Resource for development."""
        def __init__(self):
            self.resource_id = None
            self.title = None

    class ValidationResult:
        """Mock ValidationResult for development."""
        pass

    class BuilderStats:
        """Mock BuilderStats for development."""
        pass

    class BuildResult:
        """Mock BuildResult for development."""
        pass

    class PresetInfo:
        """Mock PresetInfo for development."""
        pass

    class FidelityOptions:
        """Mock FidelityOptions for development."""
        pass

    class VerificationResult:
        """Mock VerificationResult for development."""
        pass

    class ValidationRule:
        """Mock ValidationRule for development."""
        pass

    def batch_build(requests):
        """Mock batch_build for development."""
        return [f"<mock>Built from request {i}</mock>" for i, _ in enumerate(requests)]

    def validate_structure(xml):
        """Mock validate_structure for development."""
        return {"valid": True, "errors": [], "warnings": []}

# Public API
__all__ = [
    "__version__",
    "DdexBuilder",
    "Release",
    "Resource",
    "ValidationResult",
    "BuilderStats",
    "BuildResult",
    "PresetInfo",
    "FidelityOptions",
    "VerificationResult",
    "ValidationRule",
    "batch_build",
    "validate_structure",
]

# Helper function to check if using native implementation
def is_rust_available():
    """Check if the native Rust implementation is available."""
    return _RUST_AVAILABLE