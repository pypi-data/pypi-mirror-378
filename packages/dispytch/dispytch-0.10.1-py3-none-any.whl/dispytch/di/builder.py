from functools import lru_cache
from typing import Callable, Any

from dispytch.di.exc import CyclicDependencyError
from dispytch.di.extractor import extract_dependencies

from dispytch.di.tree import DependencyNode, ChildNode, DependencyTree


@lru_cache
def get_dependency_tree(func: Callable[..., Any]) -> DependencyTree:
    return DependencyTree(_build_dependency_branches(func, {}, set()))


def _build_dependency_branches(func: Callable[..., Any],
                               resolved: dict[int, DependencyNode],
                               resolving: set[int]) -> list[ChildNode]:
    children = []

    if not (dependencies := extract_dependencies(func)):
        return children

    for param_name, dependency in dependencies.items():
        if dependency.use_cache and hash(dependency) in resolved:
            children.append(ChildNode(param_name, resolved[hash(dependency)]))
            continue

        if hash(dependency) in resolving:
            raise CyclicDependencyError(f"Dependency cycle detected: {dependency}")

        resolving.add(hash(dependency))
        current_node = DependencyNode(dependency, _build_dependency_branches(dependency.func, resolved, resolving))
        resolving.remove(hash(dependency))

        if dependency.use_cache:
            resolved[hash(dependency)] = current_node

        children.append(ChildNode(param_name, current_node))

    return children
