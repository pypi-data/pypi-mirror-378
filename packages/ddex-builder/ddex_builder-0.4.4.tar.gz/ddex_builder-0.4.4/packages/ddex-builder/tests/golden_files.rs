//! Golden file tests for DDEX Builder

use ddex_builder::builder::{
    IdStrategy, LocalizedStringRequest, MessageHeaderRequest, PartyRequest, ReleaseRequest,
    TrackRequest,
};
use ddex_builder::{BuildOptions, BuildRequest, DDEXBuilder};
use insta::assert_snapshot;

#[test]
fn test_audio_album_golden() {
    let builder = DDEXBuilder::new();

    let request = BuildRequest {
        header: MessageHeaderRequest {
            message_id: None,
            message_sender: PartyRequest {
                party_name: vec![LocalizedStringRequest {
                    text: "Test Sender".to_string(),
                    language_code: None,
                }],
                party_id: None,
                party_reference: None,
            },
            message_recipient: PartyRequest {
                party_name: vec![LocalizedStringRequest {
                    text: "Test Recipient".to_string(),
                    language_code: None,
                }],
                party_id: None,
                party_reference: None,
            },
            message_control_type: Some("LiveMessage".to_string()),
            message_created_date_time: Some("2025-01-01T00:00:00Z".to_string()), // Fixed timestamp
        },
        version: "4.3".to_string(),
        profile: Some("AudioAlbum".to_string()),
        releases: vec![ReleaseRequest {
            release_id: "REL001".to_string(),
            release_reference: Some("R1".to_string()),
            title: vec![LocalizedStringRequest {
                text: "Test Album".to_string(),
                language_code: Some("en".to_string()),
            }],
            artist: "Test Artist".to_string(),
            label: Some("Test Label".to_string()),
            release_date: Some("2024-01-01".to_string()),
            upc: Some("123456789014".to_string()),
            tracks: vec![
                TrackRequest {
                    track_id: "TRK001".to_string(),
                    resource_reference: Some("A1".to_string()),
                    isrc: "USRC12345678".to_string(),
                    title: "Track One".to_string(),
                    duration: "PT3M45S".to_string(),
                    artist: "Test Artist".to_string(),
                },
                TrackRequest {
                    track_id: "TRK002".to_string(),
                    resource_reference: Some("A2".to_string()),
                    isrc: "USRC12345679".to_string(),
                    title: "Track Two".to_string(),
                    duration: "PT4M20S".to_string(),
                    artist: "Test Artist feat. Guest".to_string(),
                },
            ],
            resource_references: None,
        }],
        deals: vec![],
        extensions: None,
    };

    let options = BuildOptions {
        determinism: Some(ddex_builder::DeterminismConfig::default()),
        preflight_level: ddex_builder::preflight::PreflightLevel::Warn,
        id_strategy: IdStrategy::StableHash,
        stable_hash_config: None,
    };

    let result = builder.build(request, options).unwrap();

    // Assert against golden file
    assert_snapshot!("audio_album_output", result.xml);
    assert!(result.errors.is_empty());
    assert!(result.canonical_hash.is_some());
}

#[test]
fn test_deterministic_generation() {
    let builder = DDEXBuilder::new();

    let request = create_test_request();
    let options = BuildOptions {
        determinism: Some(ddex_builder::DeterminismConfig {
            canon_mode: ddex_builder::determinism::CanonMode::DbC14n,
            sort_strategy: ddex_builder::determinism::SortStrategy::Canonical,
            output_mode: ddex_builder::determinism::OutputMode::DbC14n,
            ..Default::default()
        }),
        preflight_level: ddex_builder::preflight::PreflightLevel::Strict,
        id_strategy: IdStrategy::StableHash,
        stable_hash_config: None,
    };

    // Generate multiple times
    let result1 = builder.build(request.clone(), options.clone()).unwrap();
    let result2 = builder.build(request.clone(), options.clone()).unwrap();
    let result3 = builder.build(request, options).unwrap();

    // All should be identical
    assert_eq!(result1.xml, result2.xml);
    assert_eq!(result2.xml, result3.xml);
    assert_eq!(result1.canonical_hash, result2.canonical_hash);
    assert_eq!(result2.canonical_hash, result3.canonical_hash);
}

#[test]
fn test_preflight_validation() {
    let builder = DDEXBuilder::new();

    // Invalid ISRC
    let mut request = create_test_request();
    request.releases[0].tracks[0].isrc = "INVALID".to_string();

    let options = BuildOptions {
        determinism: None,
        preflight_level: ddex_builder::preflight::PreflightLevel::Strict,
        id_strategy: IdStrategy::UUID,
        stable_hash_config: None,
    };

    let result = builder.build(request, options);
    assert!(result.is_err());
}

#[test]
fn test_stable_hash_ids() {
    use ddex_builder::id_generator::{HashAlgorithm, StableHashConfig, StableHashGenerator};

    let config = StableHashConfig {
        recipe: "v1".to_string(),
        algorithm: HashAlgorithm::Blake3,
        use_cache: true,
        salt: Some("test_salt".to_string()),
    };

    let mut generator = StableHashGenerator::new(config);

    // Generate IDs
    let id1 = generator
        .generate_release_id(
            "123456789012",
            "Album",
            &["USRC12345678".to_string()],
            &["US".to_string()],
        )
        .unwrap();

    // Same inputs should produce same ID
    let id2 = generator
        .generate_release_id(
            "123456789012",
            "Album",
            &["USRC12345678".to_string()],
            &["US".to_string()],
        )
        .unwrap();

    assert_eq!(id1, id2);
    assert!(id1.starts_with("B3:"));
}

#[test]
fn test_profile_validation() {
    use ddex_builder::preflight::{PreflightValidator, ValidationConfig};

    let config = ValidationConfig {
        level: ddex_builder::preflight::PreflightLevel::Strict,
        validate_identifiers: true,
        validate_checksums: true,
        check_required_fields: true,
        validate_dates: true,
        validate_references: true,
        profile: Some("AudioAlbum".to_string()),
    };

    let validator = PreflightValidator::new(config);

    // Single track should warn for AudioAlbum
    let mut request = create_test_request();
    request.releases[0].tracks = vec![request.releases[0].tracks[0].clone()];

    let result = validator.validate(&request).unwrap();
    assert!(!result.warnings.is_empty());
    assert!(result.warnings[0].code == "ALBUM_TRACK_COUNT");
}

fn create_test_request() -> BuildRequest {
    BuildRequest {
        header: MessageHeaderRequest {
            message_id: None,
            message_sender: PartyRequest {
                party_name: vec![LocalizedStringRequest {
                    text: "Test Sender".to_string(),
                    language_code: None,
                }],
                party_id: None,
                party_reference: None,
            },
            message_recipient: PartyRequest {
                party_name: vec![LocalizedStringRequest {
                    text: "Test Recipient".to_string(),
                    language_code: None,
                }],
                party_id: None,
                party_reference: None,
            },
            message_control_type: Some("LiveMessage".to_string()),
            message_created_date_time: Some("2025-01-01T00:00:00Z".to_string()), // Fixed timestamp
        },
        version: "4.3".to_string(),
        profile: Some("AudioAlbum".to_string()),
        releases: vec![ReleaseRequest {
            release_id: "REL001".to_string(),
            release_reference: Some("R1".to_string()),
            title: vec![LocalizedStringRequest {
                text: "Test Release".to_string(),
                language_code: Some("en".to_string()),
            }],
            artist: "Test Artist".to_string(),
            label: Some("Test Label".to_string()),
            release_date: Some("2024-01-01".to_string()),
            upc: Some("123456789014".to_string()),
            tracks: vec![
                TrackRequest {
                    track_id: "TRK001".to_string(),
                    resource_reference: Some("A1".to_string()),
                    isrc: "USRC12345678".to_string(),
                    title: "Test Track".to_string(),
                    duration: "PT3M30S".to_string(),
                    artist: "Test Artist".to_string(),
                },
                TrackRequest {
                    track_id: "TRK002".to_string(),
                    resource_reference: Some("A2".to_string()),
                    isrc: "USRC12345679".to_string(),
                    title: "Another Track".to_string(),
                    duration: "PT4M00S".to_string(),
                    artist: "Test Artist".to_string(),
                },
            ],
            resource_references: None,
        }],
        deals: vec![],
        extensions: None,
    }
}
