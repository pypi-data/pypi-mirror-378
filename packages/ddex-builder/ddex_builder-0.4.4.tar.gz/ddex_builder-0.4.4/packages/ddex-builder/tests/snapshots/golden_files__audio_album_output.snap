---
source: packages/ddex-builder/tests/golden_files.rs
expression: result.xml
---
<?xml version="1.0" encoding="UTF-8"?>
<ern:NewReleaseMessage xmlns:ern="http://ddex.net/xml/ern/43" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" MessageSchemaVersionId="ern/4.3">
  <MessageHeader>
    <MessageId>B3:384b9efb8b851a20399cb9d929cc204e56cac0b2e26cc8aa3dd96b89c1772cb7</MessageId>
    <MessageCreatedDateTime>2025-01-01T00:00:00Z</MessageCreatedDateTime>
    <MessageSender>
      <PartyName>Test Sender</PartyName>
    </MessageSender>
    <MessageRecipient>
      <PartyName>Test Recipient</PartyName>
    </MessageRecipient>
    <MessageThreadId>B3:384b9efb8b851a20399cb9d929cc204e56cac0b2e26cc8aa3dd96b89c1772cb7</MessageThreadId>
  </MessageHeader>
  <ResourceList>
    <SoundRecording>
      <ReferenceTitle>
        <TitleText>Track One</TitleText>
      </ReferenceTitle>
      <Duration>PT3M45S</Duration>
      <ResourceId>
        <ISRC>USRC12345678</ISRC>
      </ResourceId>
      <ResourceReference>A1</ResourceReference>
    </SoundRecording>
    <SoundRecording>
      <ReferenceTitle>
        <TitleText>Track Two</TitleText>
      </ReferenceTitle>
      <Duration>PT4M20S</Duration>
      <ResourceId>
        <ISRC>USRC12345679</ISRC>
      </ResourceId>
      <ResourceReference>A2</ResourceReference>
    </SoundRecording>
  </ResourceList>
  <ReleaseList>
    <Release>
      <ReleaseReference>R1</ReleaseReference>
      <ReleaseId>
        <GRid>REL001</GRid>
      </ReleaseId>
      <ReleaseId>
        <ICPN>123456789014</ICPN>
      </ReleaseId>
      <ReferenceTitle>
        <TitleText LanguageAndScriptCode="en">Test Album</TitleText>
      </ReferenceTitle>
      <DisplayArtistName>
        <FullName>Test Artist</FullName>
      </DisplayArtistName>
      <LabelName>
        <LabelName>Test Label</LabelName>
      </LabelName>
      <ReleaseDate>2024-01-01</ReleaseDate>
      <ReleaseResourceReference>A1</ReleaseResourceReference>
      <ReleaseResourceReference>A2</ReleaseResourceReference>
    </Release>
  </ReleaseList>
</ern:NewReleaseMessage>
